Fix issues with the form state when navigating between collection settings pages

paescuj opened the pull request:
## Scope

What's changed:

- **Switch back to `useItem` but re-initialize it on page changes**
  
  This page/component is a bit of a special case, because it is re-used with different props (collection) without being re-mounted (and thus not re-initialized) when navigating via provided links on relational fields or history.
  
  This reverts #20912 as we still want to rely on `useItem` to fetch the latest state of a collection incl. translation (was missing after #20912) as well as having the usual vars & methods available like `loading`, `save`, ... which we can pass to `v-form`.
  
  However, we need to completely re-initialize `useItem` per collection, not just fetching the new collection on page change.

- **Account for undefined collection prop in `system-field` interface**

  As outlined in https://github.com/directus/directus/pull/20912#issuecomment-1872203138

- **Improve subpath check in edit route guard**

  Previously, it would also match paths at the same level and thus wrongly allowing to exit page with unsaved changes. This happened, for example, when navigating between two collections that have the same base name (`example` and `example2`), which means the edits would be taken over from one to the other collection.

## Potential Risks / Drawbacks

Successfully tested all scenarios:
- #19417
- #20871
- https://github.com/directus/directus/pull/20949#pullrequestreview-1801865595
- "Note" for system collections is translated again
- Edit guard works for data-model, dashboards and flows pages

## Review Notes / Questions

None

paescuj committed:
Code optimization in collection fields component

paescuj committed:
Add changeset

br41nslug commented on app/src/modules/settings/routes/data-model/fields/fields.vue:
```diff
@@ -28,7 +28,7 @@ const { info: collectionInfo } = useCollection(collection);
 const fieldsStore = useFieldsStore();
 
-const { edits, item, saving, loading, save, remove, deleting } = useItem(ref('directus_collections'), collection);
+const { edits, saving, save, remove, deleting } = useItem(ref('directus_collections'), collection, {}, false);
```
```suggestion
const { edits, saving, loading, save, remove, deleting } = useItem(ref('directus_collections'), collection, {}, false);
```

br41nslug commented on app/src/modules/settings/routes/data-model/fields/fields.vue:
```diff
@@ -141,10 +141,9 @@ function discardAndLeave() {
 			<v-form
 				v-model="edits.meta"
 				collection="directus_collections"
```
```suggestion
			<v-form
				v-model="edits.meta"
				collection="directus_collections"
				:loading="loading"
```

br41nslug reviewed (changes requested):
Found a small regression discussed in DM, on the loading property. This makes the original issue reproduceable again by:
1. open a collection with relations in the data-model
2. navigate to a related collection
3. then use the browser back button

paescuj committed:
Improve subpath check in edit route guard

Previously, it would also match paths at the same level and thus wrongly
allowing to exit page. For example with two collections having the same base name
(example, example2).

paescuj committed:
Account for undefined collection prop

paescuj committed:
Switch back to `useItem` but re-initialize

This page/component is a bit of a special case, because it's re-used with different
props (collection) without being unmounted (and thus not re-initialized) when
navigating via provided links on relational fields or history.

We still want to rely on `useItem` to fetch the latest state of a
collection incl. translation as well as having the usual vars & methods
available like `loading`, `save`, ... which we can use with v-form.

However in this case, we need to completely re-initialize `useItem` per
collection, not just fetching the new collection on page change.

paescuj committed:
Update changeset

br41nslug commented on app/src/modules/settings/routes/data-model/fields/fields.vue:
```diff
@@ -24,15 +23,16 @@ const { t } = useI18n();
 
-const { edits, item, saving, loading, save, remove, deleting } = useItem(ref('directus_collections'), collection);
+// Re-initialize if collection (page) changes
+const usableItem = computed(() => useItem(ref('directus_collections'), collection));
```
I dont think wrapping the already reactive return values of `useItem` in an extra computed is a better approach than the previous one. Why was this entire PR rewritten after a small piece of feedback with an easy fix? 🤔

br41nslug replied:
Should we not fix this at the source then in `useItems` so it properly resets its internal state when the input refs `collection`, `primaryKey` or `query` change? Putting `useItem` in a computed watching the same variables as it does internally feels like treating the symptoms too 😬

paescuj replied:
Fair point! I hesitated to directly adjust `useItem` because it is used in several places and in different ways, but I took a closer look now and found no problem with that.

br41nslug replied:
Very understandable, I was indeed afraid that changing `useItem` would be regression/bug prone but it did sound like the right place to fix it so glad you took a closer look 😄

br41nslug reviewed (commented):
Feels like this PR has experienced some scope creep after my last review 😬

paescuj committed:
Full reset in useItems on refresh

paescuj committed:
Don't clear edits

Should be checked & cleared externally, e.g. via edit guard

paescuj committed:
Missed to remove `ìmmediate` option

br41nslug committed:
Merge branch 'main' into optimize-collection-fields-edit

br41nslug reviewed (approved):
Great work! Looks like this finally tackles the underlying cause 😄
